scripts/emacs_lua_complete.lua

```diff
@@ -0,0 +1,79 @@
+--    __emacs_lua_complete:
+-- This function finds and prints a limited number of matches on the
+-- lua globals tree, _G. See `lua-complete-string`. N.B.: Does not
+-- complete numeric keys (e.g. for arrays) or keys which begin with an
+-- underscore.
+-- Arguments:
+--   PARTS: a table array of dot-separated parts of the completion item
+--     (e.g. table.con would become {"table","con"}).  All but the last
+--     element must be valid subkeys in the _G heirarchy.
+--   LIBS: an array of tables of the form
+--       {var: varname, lib: libfile}
+--     for e.g. local varname=require("libfile"). This library will be
+--     require'd and searched alongside global variables.
+--   LOCALS: an array of strings parsed from local aloc,bloc = ... to
+--     complete as well
+--
+-- Example:
+-- __emacs_lua_complete({'table','ins'},{{var="xyz",lib="libfile"}},
+--                      {"frank","beans"},5000)
+--
+-- Written 2019/05 J.D. Smith
+function __emacs_lua_complete(parts, libs, locals, limit)
+   -- Setup a FIFO queue for breadth-first tree traversal
+   local queue = {head = 0, tail = 0}
+   function queue:push(value)
+      self.tail = self.tail+1  self[self.tail] = value
+   end
+   function queue:pop()
+      self.head = self.head+1  return self[self.head]
+   end
+
+   -- Mirror the globals table
+   local globals,cnt,limit = {}, 0, limit or 2000
+   for k,v in pairs(_G) do globals[k] = v end
+   
+   -- Add libs and locals
+   if libs then
+      for _,v in ipairs(libs) do
+	 local good,lib=pcall(require,v.lib)
+	 globals[v.var] = good and lib or {}
+      end
+   end
+   if locals then
+      for _,v in ipairs(locals) do globals[v] = {} end
+   end
+
+   -- Descend the tree to the limit of dotted parts
+   local g,pre = globals, ""
+   for i = 1,#parts do
+      if not g then break end	-- should always exist
+      if i < #parts then -- keep descending
+	 pre = (i > 1 and pre.."." or "") .. parts[i]
+	 g = g[parts[i]]
+      else  -- final part; seed the queue with any matching tables
+	 for k,v in pairs(g) do
+	    if type(k) == "string" and k:sub(1,1) ~= "_" and k:find('^'..parts[i]) then
+	       local dpath=(i > 1 and pre .. "." or "") .. k
+	       print(dpath)
+	       cnt = cnt + 1
+	       if type(v) == "table" then queue:push({pre = dpath,entries = v}) end
+	    end
+	 end
+      end
+   end
+
+   -- Perform a limited breadth-first traversal of the queued tables
+   while queue.head < queue.tail and cnt < limit do
+      local q = queue:pop()
+      for k,v in pairs(q.entries) do
+	 if type(k) == "string" and k:sub(1,1) ~= "_" then 
+	    pre = q.pre .. "." .. k
+	    print(pre)
+	    cnt = cnt + 1
+	    if cnt >= limit then break end
+	    if type(v) == "table" then queue:push({pre = pre,entries = v}) end
+	 end
+      end
+   end
```

Comment:
Wait, I just realized, this code traversing the tables beyond the initial completion, e.g. you request
`foo.b` and with `foo.bar` being a valid completion option, it would also go ahead and complete it as `foo.bar.baz`, `foo.bar.qux` and so on, right? If my understanding is correct, at the risk of sounding rude when the implementation has already come this far, I'd say that's too much to my liking. I am accustomed to step-by-step completions, e.g. when you complete one level, input dot and complete the next one. Is there a performance reason to fetch completions in batches of 1000s?

And just out of curiosity, do you have a precedent for this kind of behaviour in other Emacs modes to see if maybe I am lagging behind the Emacs innovation frontier?

Reply:
Yes, and this took some real doing, it's the whole point of the breadth-first search.  This is the strongly recommended approach to completion by the completion-at-point facility.  The reason is that there are various completion "styles", and the completion backend is supposed to be agnostic about them, returning a "full list" and letting the completion style sort out what it wants.  For example, [initials](https://www.gnu.org/software/emacs/manual/html_node/emacs/Completion-Styles.html).  On that topic the `completion-at-point-functions` docs include:

> NOTE: These functions should be cheap to run since they’re sometimes
> run from ‘post-command-hook’; and they should ideally only choose
> which kind of completion table to use, and not pre-filter it based
> on the current text between START and END (e.g., they should not
> obey ‘completion-styles’).

and in the manual:

> This specifies an all-completions operation. The function should return a list of all possible completions of the specified string. 

The advantage here is that you can prefetch the deep tree (which is often only dozens deep), and then it is cached for subsequent completes, and all the tests and other calls the completion framework make.  It is sorted so that the "first after dot" is up front, so basically you can ignore it.  It _radically_ cuts down on the back and forth traffic to the lua shell.  

Have you tried it out?  Effectively it does the same thing: complete to the next dot (unless there is only ONE completion beyond that dot, which would be rare).  But far more efficiently and with the option to try some crazy completion styles with company mode, etc.

Reply:
It is not that I am particularly against "deep" completions, although admittedly I prefer the shorter version and no craziness. 

It's just that we already do completions on a live interpreter that could produce some side-effects, we have to be careful about it. With "deep" completions it takes one object hidden somewhere within the initial 1000 element set to have a `__pairs` method doing something crazy to have a really bad experience with this. And I'm don't mean malevolent, just crazy. With "short" completions I as a user have some control over whose keys I want to be iterated in the background and whose not. With "deep" completions I am at the mercy of the code of the library, or I have to verify each library with my own eyes.

> It radically cuts down on the back and forth traffic to the lua shell.

Many Pandora boxes have been opened with good intentions to optimize things. Spectre/Meltdown family of CPU vulnerabilities immediately come to mind, but surely there are more.

Reply:
Hmm, I hadn't thought of rogue objects overriding __pairs.  Can you mock up an example?  Maybe there's a way to protect against that.  Deep completions are now a fundamental part of modern Emacs completion framework, for better or worse.  Given a starting fragment, you are ideally to return all possible completions of that fragment without prejudice, which enables many modern forms of completion that many people use.  If this is a deal-breaker for you, I really do wish you had voiced the concerns back when I first proposed a depth-first search of the globals tree, and before I spent ~days implementing it.

Reply:
If you don't mind, I'll respond to this as a generic comment, as this is a quite deep topic (no pun intended) and it becomes increasingly difficult to keep track of all the "line comment threads" as lines are updated and the comments themselves become outdated.